zebrad/src/commands/start/sync.rs

```diff
@@ -0,0 +1,190 @@
+                    self.request_blocks(hashes).await?;
+                }
+                Ok(_) => continue,
+                Err(e) => {
+                    error!("{:?}", e);
+                }
+            }
+        }
+
+        for tip in tip_set {
+            let mut block_locator = self.block_locator.clone();
+            block_locator[0] = tip;
+            self.populate_prospectives(block_locator).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Queue downloads for each block that isn't currently known to our node
+    pub async fn request_blocks(&mut self, mut hashes: Vec<BlockHeaderHash>) -> Result<(), Report> {
+        hashes.retain(|hash| !self.known_block(hash));
+
+        for chunk in hashes.chunks(10usize) {
+            self.queue_download(chunk).await?;
+        }
+        Ok(())
+    }
+
+    /// Drive block downloading futures to completion and dispatch downloaded
+    /// blocks to the validator
+    pub async fn process_blocks(&mut self) -> Result<(), Report> {
+        while let Some(res) = self.block_requests.next().await {
+            match res.map_err::<Report, _>(|e| eyre!(e)) {
+                Ok(zebra_network::Response::Blocks(blocks)) => {
+                    for block in blocks {
+                        assert!(
+                            self.downloading.remove(&block.as_ref().into()),
+                            "all received blocks should be explicitly requested and received once"
+                        );
+                        self.validate_block(block).await?;
+                    }
+                }
+                Ok(_) => continue,
+                Err(e) => {
+                    error!("{:?}", e);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Validate a downloaded block using the validator service, inserting the
+    /// block into the state if successful
+    #[tracing::instrument(skip(self))]
+    async fn validate_block(
+        &mut self,
+        block: std::sync::Arc<zebra_chain::block::Block>,
+    ) -> Result<(), Report> {
+        let fut = self
+            .state
+            .ready_and()
+            .await
+            .map_err(|e| eyre!(e))?
+            .call(zebra_state::Request::AddBlock { block });
+
+        let _handle = tokio::spawn(
+            async move {
+                match fut.await.map_err::<Report, _>(|e| eyre!(e)) {
+                    Ok(_) => {}
+                    Err(report) => error!("{:?}", report),
+                }
+            }
+            .in_current_span(),
+        );
+
+        Ok(())
+    }
+
+    /// Returns true if the block is being downloaded or has been validated
+    fn known_block(&self, hash: &BlockHeaderHash) -> bool {
+        self.downloading.contains(hash) || self.state_contains(hash)
+    }
+
+    /// Returns true if the block has been validated and inserted into the chain
+    /// state
+    fn state_contains(&self, hash: &BlockHeaderHash) -> bool {
+        todo!()
```

Comment:
@ZcashFoundation/zebra-team We need to figure out how to handle this. The options as I see them are either:

* return a concrete type from `init` that has a `fn contains` for querying the sled layer
* Add a variant to the `Request`/`Response` types for checking if blocks are present in the state
* create an extension trait for `trait StateService: Service` that just adds a `fn contains` and return that instead of an `impl Service`

Reply:
I think the cleanest option would be to add a variant to the state request type to query whether some block hash is contained in the current best chain, since this is probably useful in other contexts too.  I'm not sure what the response should be exactly, but I think that the most useful thing would be to return the number of blocks above the given block in the best chain, since this is also useful for deciding on transaction confirmations.

Reply:
When we were doing gardening last week @teor2345 brought up the suggestion of just storing a list of all downloaded block hashes within the downloader as a simple and reliable initial solution to this problem, so I ended up taking their approach instead of adding the `Contains` request. 

Instead I added a `GetKnownBlockHashes` request that could be made when constructing the downloader by asking the state layer of all blocks it already knows about, which can be used as a starting point for the list of known blocks that we shouldn't bother downloading, and then from there we just insert each new hash that we download as we go.

Reply:
What happens if a block is temporarily unavailable?
For example, it's a new block that only a few peers know about.

What happens if a block is temporarily unverifiable?
For example, its miner time is too far in the future.

In either case, we'd want to retry these block downloads after a timeout.